ngraph/core/reference/include/ngraph/runtime/reference/einsum.hpp

```diff
@@ -0,0 +1,1210 @@
+                    std::string required_subscript = "";
+                    for (auto index : labels_inds)
+                    {
+                        required_subscript += labels[index];
+                    }
+                    is_separate_first = true;
+                    return required_subscript;
+                }
+
+                std::unordered_map<std::string, std::vector<size_t>>
+                    compute_label_dim_map(const Rank& input_rank,
+                                          const std::string& input_subscript)
+                {
+                    const std::string ellipsis = "...";
+                    auto labels = ngraph::opset7::Einsum::extract_labels(input_subscript);
+                    size_t input_rank_length = labels.size();
+                    if (input_rank.is_dynamic() &&
+                        std::find(labels.begin(), labels.end(), ellipsis) != labels.end())
+                    {
+                        NGRAPH_CHECK(
+                            false,
+                            "Input rank cannot be dynamic in case of ellipsis in input subscript");
+                    }
+                    else if (input_rank.is_static())
+                    {
+                        input_rank_length = input_rank.get_length();
+                    }
+                    std::unordered_map<std::string, std::vector<size_t>> resulted_map;
+                    NGRAPH_CHECK(input_rank_length >= labels.size());
+                    size_t num_broadcasted_dims = input_rank_length - labels.size() + 1;
+                    NGRAPH_CHECK(num_broadcasted_dims >= 0);
+
+                    size_t current_dim = 0;
+                    for (const auto& label : labels)
+                    {
+                        if (label == ellipsis)
+                        {
+                            std::vector<size_t> label_dims;
+                            for (size_t ind = 0; ind < num_broadcasted_dims; ++ind)
+                            {
+                                label_dims.push_back(current_dim + ind);
+                            }
+                            resulted_map[label] = label_dims;
+                            current_dim += num_broadcasted_dims;
+                        }
+                        else if (resulted_map.find(label) != resulted_map.end())
+                        {
+                            resulted_map[label].push_back(current_dim);
+                            ++current_dim;
+                        }
+                        else
+                        {
+                            std::vector<size_t> label_dims;
+                            label_dims.push_back(current_dim);
+                            resulted_map[label] = label_dims;
+                            ++current_dim;
+                        }
+                    }
+
+                    return resulted_map;
+                }
+
+                Shape compute_sub_shape(const Shape& input_shape,
```

Comment:
Is it possible to reuse `ngraph::shape` routine here?
https://github.com/openvinotoolkit/openvino/blob/master/ngraph/core/include/ngraph/shape.hpp#L23

Example of using is here https://github.com/openvinotoolkit/openvino/blob/master/ngraph/core/reference/include/ngraph/runtime/reference/gather.hpp#L32-L33

Reply:
I did changes inside this function. I only managed to use `shape_size` function. I can't assign `Shape` object with `Span` value.